openmsupply/application-manager-server
# #85 Actions for Change Application Stage and Status

Created Actions "incrementStage" and "changeStatus", and inserted them into existing templates. Note that "incrementStage" automatically creates an associated "Draft" status record if none exists, so when creating a new application, only one call of "incrementAction" is required.

Also addressed the following as part of this PR:
- created Postgres functions to ensure that there can only ever be one `is_current = true` per stage and status: whenever `is_current` is set to `true`, the function automatically sets `is_current = false` for all other status or stage records for that application. (The plugin(s) also assumes that any new status or stage record will be the new `is_current` record, so that's applied automatically for every update)
- a few additional PostgresDB methods for getting and setting application statuses and stages.
- started a shared `types` file for plugins.

Also, I really don't understand the need for the adding the "Completed" status_history record when incrementing the Stage. It's never `is_current`, and it doesn't provide any additional information (and it's not a real record of any history). It'd be good if reviewer could suggest a situation/query where it could be at all helpful, because it just seems to add unnecessary confusion to me and serves no purpose. (Ideally I'd get rid of lines 50-58 of `incrementStage.ts`) 

**To test:**
In order to test from the front-end, you'll need my branch that adds the "Create" trigger. PR [here](https://github.com/openmsupply/application-manager-web-app/pull/79)
- reset the database and reload the server.
- Proceed to start a new user registration application from the web browser front-end.
- You should see a new application with Status "Draft" and Stage "Automatic" in the database (see the `application_stage_status` VIEW to see in combined view)
- Complete and Submit the application.
- The application should now be set to "Status: Completed" (and Outcome: Approved). Note: The user information is NOT added to the database correctly, as the current input plugin is not sending the response as correctly structured JSON, but the `createUserFromApplication` Action is working fine when responses are manually input.

To check other possible stage/status updates, use manual SQL queries against the other templates/applications:
- the Company Registration template has two stages, which you can test by updating the Trigger in the partial Company Registration application.
   - "onApplicationCreate" increments the stage (and creates new "Draft" status). You can just run this trigger multiple times to see it increment through all the stages -- not realistic, but demonstrates the plugin working.
   - "onApplicationSubmit" changes the status to "Submitted".
   - "onReviewSave" trigger (on Review table) also increments the stage.
